api/versions.go

```diff
@@ -465,20 +466,31 @@ func (api *DatasetAPI) publishVersion(ctx context.Context, currentDataset *model
 		}
 
 		// Pass in newVersion variable to include relevant data needed for update on dataset API (e.g. links)
-		if err := api.publishDataset(ctx, currentDataset, versionDoc); err != nil {
+		if err := api.publishDataset(ctx, currentDataset, versionUpdate); err != nil {
 			log.Error(ctx, "putVersion endpoint: failed to update dataset document once version state changes to publish", err, data)
 			return err
 		}
 
 		// Only want to generate downloads again if there is no public link available
 		if currentVersion.Downloads != nil && currentVersion.Downloads.CSV != nil && currentVersion.Downloads.CSV.Public == "" {
-			if err := api.downloadGenerator.Generate(ctx, versionDetails.datasetID, versionDoc.ID, versionDetails.edition, versionDetails.version); err != nil {
-				data["instance_id"] = versionDoc.ID
-				data["state"] = versionDoc.State
+			// Lookup the download generator using the version document type
+			t, err := models.GetDatasetType(currentVersion.Type)
+			if err != nil {
+				return fmt.Errorf("error getting type of version: %w", err)
+			}
+			generator, ok := api.downloadGenerators[t]
+			if !ok {
+				return fmt.Errorf("no downloader available for type %s", t.String())
+			}
+			// Send Kafka message.  The generator which is used depends on the type defined in VersionDoc.
+			if err := generator.Generate(ctx, versionDetails.datasetID, versionUpdate.ID, versionDetails.edition, versionDetails.version); err != nil{
+				data["instance_id"] = versionUpdate.ID
+				data["state"] = versionUpdate.State
 				log.Error(ctx, "putVersion endpoint: error while attempting to generate full dataset version downloads on version publish", err, data)
-				// TODO - TECH DEBT - need to add an error event for this.
 				return err
+				// TODO - TECH DEBT - need to add an error event for this.  Kafka message perhaps.
 			}
+			log.Info(ctx, "putVersion endpoint: generated full dataset version downloads for type %s", data)
```

Comment:
I would suggest adding the type to the `data`
  ie.
  data["type"] = t.String()